### PR TITLE
[WIP] eth_getTransactionReceipt to return optional status field

### LINKED ISSUE #407: Detect failed via eth_getTransactionReceipt
Hi, I have a question regarding detection of failed transactions (out of gas error) in Geth Classic. In last fork Ethereum introduced change in protocol allowing `eth_getTransactionReceipt` to return additional `status` field which can be used for validation. However, this flag is not present in Ethereum Classic protocol. How can I detect that incoming contract transaction was failed or not? I was using a trick with marking as failed each transaction that used all of its gas, however its not 100% accurate. I see the block explorers somehow can distinguish this. Is there a solution for this in Classic yet I dont know of?

### PR DESCRIPTION
**Pull request by whilei**
See https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_gettransactionreceipt

Implements `receipt.Status`, adding return bool value 'failed' to core.ApplyMessage,
and refactoring core.TransitionDb to handle this. 

Rel and fixes #407 

- [ ] update tests

### CODE REVIEW

File: core/types/receipt.go [hunk 1]
```diff
@@ -26,58 +27,112 @@ import (
...
+	}
+	return r
 }
 
 // EncodeRLP implements rlp.Encoder, and flattens the consensus fields of a receipt
-// into an RLP stream.
+// into an RLP stream. If no post state is present, byzantium fork is assumed.
 func (r *Receipt) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{r.PostState, r.CumulativeGasUsed, r.Bloom, r.Logs})
+	return rlp.Encode(w, &receiptRLP{r.statusEncoding(), r.CumulativeGasUsed, r.Bloom, r.Logs})
```
Comment (sorpaas) [inline core/types/receipt.go]: Looks like this hard fork?

**Comment by whilei** [comment]
@sorpaas Good catch, thank you. I agree with you that it seems modifying ReceiptStorage, and leaving consensus receipt unchanged, seems preferable. 

A question:
> However, even in the later case I would still worry compatibility with Parity (which implements the EIP-658 hard fork for Foundation's chain).

If Parity already implements the HF change, can you explain more what your concern is about compatibility? If without the change we're already incompatible with them since they've already adopted the change? If the storage change would only impact the API, and it would then match(?) EF and Parity APIs?

**Comment by whilei** [comment]
Also, what is NACK?

**Comment by sorpaas** [comment]
@whilei I use NACK for negative acknowledgement, and "Concept NACK" if I don't agree with the direction.

Currently we're compatible with Parity. However, adding this means we have one more field return value than them. That may cause problem for dapp developers.

**Comment by sorpaas** [comment]
@whilei The `eth_` namespace has caused some problem in the past, as well. For example, `eth_chainId`. My suggestion is as follows:

1. In stateDB, implement the `receiptStorage` change, and **double check that you left everything else unchanged in case it causes hard fork**.
2. Leave `eth_getTransactionReceipt` call unchanged. This keeps compatibility with Parity.
3. Create another RPC call `etc_getTransactionReceipt`. This is the same as `eth_getTransactionReceipt` but include `status` field **whenever available from genesis block**.

This requires a little bit more coding. There needs to be customized logic to check whether a receipt contains status field and cannot just use a hard fork number. What's stored in the storage now is also different from block's `receiptRoot`, so there'll also need to be two RLP encoding scheme and we'll need to check the whole codebase to use the correct one. However, it is much more sophisticated compared with EIP-658 and needs no hard fork.

Together with this, I would suggest to change the name of `eth_chainId` to `etc_chainId`, so we have a potential to agree on an RPC call name on this with Parity, as well.

**Comment by whilei** [comment]
For reference, https://github.com/paritytech/parity/blob/28c731881f2da0ceca4752dbcc1a8f9ad041f988/ethcore/types/src/receipt.rs#L62-L108

**Comment by sorpaas** [comment]
@whilei EIP658 is a hard fork because it changes the receipt RLP encoding.

**Comment by tomokazukozuma** [comment]
When would this be added?

**Comment by yourihua** [comment]
When would this be added?

**Comment by tzdybal** [comment]
I'm closing this PR because it's replaced by #634. 

@tomokazukozuma, @yourihua - this should be included in next release (hopefully before end of the week).

**Pull request closed without merging**